dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ResourcesServiceImpl.java

```diff
@@ -17,1286 +17,404 @@
-        // String directoryName = storageOperate.getFileName(type, tenantCode, fullName);
-        String resourceRootPath = storageOperate.getDir(type, tenantCode);
+    @Override
+    public void createFile(CreateFileRequest createFileRequest) {
+        CreateFileDto createFileDto = createFileRequestTransformer.transform(createFileRequest);
+        createFileDtoValidator.validate(createFileDto);
+
+        // todo: use storage proxy
+        MultipartFile file = createFileDto.getFile();
+        String fileAbsolutePath = createFileDto.getFileAbsolutePath();
+        String srcLocalTmpFileAbsolutePath = copyFileToLocal(file);
         try {
-            if (!storageOperate.exists(resourceRootPath)) {
-                storageOperate.createTenantDirIfNotExists(tenantCode);
-            }
-
-            if (!storageOperate.mkdir(tenantCode, fullName)) {
-                throw new ServiceException(String.format("Create resource directory: %s failed.", fullName));
-            }
-            putMsg(result, Status.SUCCESS);
-        } catch (Exception e) {
-            throw new ServiceException(String.format("create resource directory: %s failed.", fullName));
+            storageOperator.upload(srcLocalTmpFileAbsolutePath, fileAbsolutePath, true, false);
+            ApiServerMetrics.recordApiResourceUploadSize(file.getSize());
+            log.info("Success upload resource file: {} complete.", fileAbsolutePath);
+        } catch (Exception ex) {
+            // If exception, clear the tmp path
+            FileUtils.deleteFile(srcLocalTmpFileAbsolutePath);
+            throw ex;
         }
     }
 
-    /**
-     * upload file to hdfs
-     *
-     * @param loginUser login user
-     * @param fullName  full name
-     * @param file      file
-     * @param type      resource type
-     * @return upload success return true, otherwise false
-     */
-    private boolean upload(User loginUser, String fullName, MultipartFile file, ResourceType type) {
-        // save to local
-        String fileSuffix = Files.getFileExtension(file.getOriginalFilename());
-        String nameSuffix = Files.getFileExtension(fullName);
-
-        // determine file suffix
-        if (!fileSuffix.equalsIgnoreCase(nameSuffix)) {
-            return false;
-        }
-        // query tenant
-        String tenantCode = getTenantCode(loginUser);
-        // random file name
-        String localFilename = FileUtils.getUploadFilename(tenantCode, UUID.randomUUID().toString());
-
-        // save file to hdfs, and delete original file
-        String resourcePath = storageOperate.getDir(type, tenantCode);
+    @Override
+    public void createFileFromContent(CreateFileFromContentRequest createFileFromContentRequest) {
+        CreateFileFromContentDto createFileFromContentDto =
+                createFileFromContentRequestTransformer.transform(createFileFromContentRequest);
+        createFileFromContentDtoValidator.validate(createFileFromContentDto);
```

Comment:
DTO validation logic is used in many places. How about using AOP to reduce duplicate code through annotations?

Reply:
May be we can not use apo, since some case will transform and validate many times. This is not a huge duplicate.